File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.maxi = 0
        self.total = 0

    def maxProduct(self, root: Optional[TreeNode]) -> int:
        self.dfs_total(root)

        self.dfs_max(root)

        # return self.maxi
        return self.maxi % (10**9 + 7)

    def dfs_total(self, node) :
        if node == None : return

        self.total += node.val
        self.dfs_total(node.left)
        self.dfs_total(node.right)

    def dfs_max(self, node) :
        if node == None : return 0

        left, right = self.dfs_max(node.left), self.dfs_max(node.right)

        sub_tree = left + right + node.val

        prod = sub_tree * (self.total - sub_tree)
        if prod > self.maxi : self.maxi = prod

        return sub_tree

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int MOD = 1000000007;
    long total = 0;
    long max = 0;
    public int maxProduct(TreeNode root) {

        // Logic for the problem ::
        dfsTotal (root);
        dfsMax (root);
        // System.out.println(total);

        return (int)(max % MOD);
    }

    private long dfsTotal (TreeNode node) {
        if (node == null) return 0;
        total += node.val;

        return total + dfsTotal(node.left) + dfsTotal(node.right);
    }

    private long dfsMax (TreeNode node) {
        if (node == null) return 0;

        // long left = dfsMax(node.left);
        // long right = dfsMax(node.right);
        long subTree = dfsMax(node.left) + dfsMax(node.right) + node.val;

        long prod = subTree * (total - subTree);
        max = Math.max(max, prod);

        return subTree;
    }
}